pkg/services/auth/authimpl/auth_token.go

```diff
@@ -504,11 +583,29 @@ func (s *UserAuthTokenService) rotatedAfterParam() int64 {
 	return getTime().Add(-s.cfg.LoginMaxInactiveLifetime).Unix()
 }
 
+func createToken() (string, error) {
+	token, err := util.RandomHex(16)
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
 func hashToken(token string) string {
 	hashBytes := sha256.Sum256([]byte(token + setting.SecretKey))
 	return hex.EncodeToString(hashBytes[:])
 }
 
+func generateAndHashToken() (string, string, error) {
+	token, err := createToken()
+	if err != nil {
+		return "", "", err
+	}
+
+	return token, hashToken(token), nil
```

Comment:
Nitpick: a [HMAC](https://golangcode.com/generate-sha256-hmac/) would be safer from a cryptographic point of view (even though [Length extension attack cannot](https://en.wikipedia.org/wiki/Length_extension_attack) happen because the random token is not a user input)

Reply:
Agreed, clearly not worth the effort